monkey/monkey_island/cc/services/agent_plugin_service/mongo_agent_plugin_repository.py

```diff
@@ -0,0 +1,245 @@
+import logging
+from enum import Enum
+from typing import Any, Dict, Optional
+
+import gridfs
+from pymongo import MongoClient
+
+from common import OperatingSystem
+from common.agent_plugins import AgentPlugin, AgentPluginManifest, AgentPluginType
+from monkey_island.cc.repositories import (
+    RemovalError,
+    RetrievalError,
+    StorageError,
+    UnknownRecordError,
+)
+
+from .i_agent_plugin_repository import IAgentPluginRepository
+
+BINARY_OS_MAPPING_KEY = "binaries"
+
+logger = logging.getLogger(__name__)
+
+
+class MongoAgentPluginRepository(IAgentPluginRepository):
+    def __init__(self, mongo_client: MongoClient) -> None:
+        self._agent_plugins_collection = mongo_client.monkey_island.agent_plugins
+        self._agent_plugins_binaries_collections = self._get_binary_collections(mongo_client)
+
+    def _get_binary_collections(
+        self, mongo_client: MongoClient
+    ) -> Dict[OperatingSystem, gridfs.GridFS]:
+        # Note: This expects the database to have a collection for each operating system
+        agent_plugins_binaries_collections: Dict[OperatingSystem, gridfs.GridFS] = {}
+
+        for os in OperatingSystem:
+            agent_plugins_binaries_collections[os] = gridfs.GridFS(
+                mongo_client.monkey_island, self._get_binary_collection_name(os)
+            )
+
+        return agent_plugins_binaries_collections
+
+    def _get_binary_collection_name(self, operating_system: OperatingSystem) -> str:
+        return f"agent_plugins_binaries_{operating_system.value}"
+
+    def get_plugin(
+        self, host_operating_system: OperatingSystem, plugin_type: AgentPluginType, name: str
+    ) -> AgentPlugin:
+        # TODO: Figure out how to store/load binaries for each operating system.
+        #       - We can store a dict field with the ID of the binary for each operating system
+        #       - We remove this field and add the
+        # - Query the agent_plugins collection for the serialized plugin
+        # - Get the binary ID for the given OS from the serialized plugin
+        # - Query the agent_plugins_binaries collection for the binary
+        # - Remove the binary mapping from the serialized plugin
+        # - Add the binary to the serialized plugin
+        # - Deserialize the plugin
+        try:
+            plugin_dict = self._get_agent_plugin(plugin_type, name)
+        except Exception:
+            raise UnknownRecordError(
+                f"Error retrieving the agent plugin {name} of type {plugin_type}"
+            )
```

Comment:
Consider moving this error handling down into `_get_agent_plugin()`.

Reply:
Hmm. We could. Should we modify `IAgentPluginRepository.remove_agent_plugin()` to indicate that it can raise an UnknownRecordError then?

Reply:
No. It can catch `UnknownRecordError` and just do nothing.